protos/table.proto

```diff
@@ -179,13 +179,46 @@ message DataFile {
   string path = 1;
   // The ids of the fields/columns in this file.
   //
-  // IDs are assigned based on position in the file, offset by the max existing
-  // field id in the table (if any already). So when a fragment is first created
-  // with one file of N columns, the field ids will be 1, 2, ..., N. If a second,
-  // fragment is created with M columns, the field ids will be N+1, N+2, ..., N+M.
+  // In Lance v1 IDs are assigned based on position in the file, offset by the max
+  // existing field id in the table (if any already). So when a fragment is first
+  // created with one file of N columns, the field ids will be 1, 2, ..., N. If a
+  // second, fragment is created with M columns, the field ids will be N+1, N+2,
+  // ..., N+M.
   //
-  // These ids must be sorted and contiguous.
+  // In Lance v1 there is one field for each field in the input schema, this includes
+  // nested fields (both struct and list).  Fixed size list fields have only a single
+  // field id (these are not considered nested fields in Lance v1).
+  //
+  // This allows column offsets to be calculated from field IDs and the input schema.
+  //
+  // In Lance v2 the field IDs generally follow the same pattern but there is no
+  // way to calculate the column offset from the field ID.  This is because a given
+  // field could be encoded in many different ways, some of which occupy a different
+  // number of columns.  For example, a struct field could be encoded into N + 1 columns
+  // or it could be encoded into a single packed column.  To determine column offsets
+  // the column_offsets property should be used instead.
+  //
+  // In Lance v2 a nested field may or may not have its field id recorded here.  In the
+  // "packed struct" example above a nested field would not have a field id recorded here.
+  // If a nested field is not listed here then it is not possible for that field to be
+  // projected / individually loaded and the entire parent field must be loaded instead (this
+  // is the downside of the packed struct)
```

Comment:
What do you think of making it so we have all `field_ids` represented, but if it's a child field that doesn't exist separately from it's parent, we have the `column_offset` be `-1`. That way we keep `field_ids` decoupled from the layout details of the file format, but the V2-specific `column_offsets` encodes that instead. WDYT?

Reply:
> Currently there is an invariant: if a field_id is in the schema, it exists in at least one data file. Is the invariant now: if a field_id is in the schema, it or its ancestor must be in one of the data files?

Yes, I'll add a comment to this effect.  We can probably be stronger too.  Each field id should have at most one match.  If an ancestor has a match then none of its children should have matches.

> What do you think of making it so we have all field_ids represented, but if it's a child field that doesn't exist separately from it's parent, we have the column_offset be -1. That way we keep field_ids decoupled from the layout details of the file format, but the V2-specific column_offsets encodes that instead. WDYT?

That seems fine.  Note, the column offsets will still have gaps because some fields may be stored as multiple columns (e.g. strings).